### PR TITLE
tracing: add support for OpenCensus

### PR DESCRIPTION
**Pull request by brianwolfe**
Add support for OpenCensus tracing. This is similar to the existing
Stackdriver support. This is tested by adding a new tracing bootstrap
golden test comparison and by individually testing the configuration
options, similar to the existing tracers.

This implements the API from this PR:
https://github.com/istio/api/pull/1563

After this lands, I will update the documentation for tracing for OpenCensusAgent and in general for context propagation, since the existing documentation is a bit confusing about which headers need to be propagated in which cases.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[X] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by brianwolfe** [comment]
@nrjpoddar, @douglas-reid PTAL.

File: pkg/bootstrap/option/convert.go [hunk 1]
```diff
@@ -180,6 +181,32 @@ func durationConverter(value *types.Duration) convertFunc {
...
+			switch c {
+			case meshAPI.Tracing_OpenCensusAgent_W3C_TRACE_CONTEXT:
+				envoyContexts = append(envoyContexts, "TRACE_CONTEXT")
+			case meshAPI.Tracing_OpenCensusAgent_GRPC_BIN:
+				envoyContexts = append(envoyContexts, "GRPC_BIN")
+			case meshAPI.Tracing_OpenCensusAgent_CLOUD_TRACE_CONTEXT:
+				envoyContexts = append(envoyContexts, "CLOUD_TRACE_CONTEXT")
+			case meshAPI.Tracing_OpenCensusAgent_B3:
+				envoyContexts = append(envoyContexts, "B3")
+				// Ignore UNSPECIFIED
```
Comment (douglas-reid) [inline pkg/bootstrap/option/convert.go:203]: super minor nit: i might move this comment up to the top above W3_TRACE_CONTEXT, as the indent makes it a bit funky at the bottom.

File: pkg/bootstrap/option/convert.go [hunk 2]
```diff
@@ -180,6 +181,32 @@ func durationConverter(value *types.Duration) convertFunc {
 	}
 }
 
+// openCensusAgentContextConverter returns a converter that returns the list of
+// distributed trace contexts to propagate with envoy.
+func openCensusAgentContextConverter(contexts []meshAPI.Tracing_OpenCensusAgent_TraceContext) convertFunc {
+	return func(*instance) (interface{}, error) {
+		if len(contexts) == 0 {
+			return convertToJSON([]string{"TRACE_CONTEXT", "GRPC_BIN", "CLOUD_TRACE_CONTEXT", "B3"}), nil
```
Comment (douglas-reid) [inline pkg/bootstrap/option/convert.go:189]: probably doesn't make a difference, but should we just have this be a preconverted const or var (something like AllContexts) ?

**Review by douglas-reid — APPROVED** [review]

File: pkg/bootstrap/option/convert.go (hunk 1 above)
Comment (brianwolfe) [reply pkg/bootstrap/option/convert.go]: thanks, that looks better now.

File: pkg/bootstrap/option/convert.go (hunk 2 above)
Comment (brianwolfe) [reply pkg/bootstrap/option/convert.go]: SG. I wanted to keep the scope of envoy trace contexts to just this function so it's easy to extract out/modify when the new telemetry api starts to land. Made it into a pre-converted var in the function.

**Comment by brianwolfe** [comment]
@howardjohn PTAL

File: pkg/bootstrap/option/convert.go [hunk 3]
```diff
@@ -180,6 +181,33 @@ func durationConverter(value *types.Duration) convertFunc {
 	}
 }
 
+// openCensusAgentContextConverter returns a converter that returns the list of
+// distributed trace contexts to propagate with envoy.
+func openCensusAgentContextConverter(contexts []meshAPI.Tracing_OpenCensusAgent_TraceContext) convertFunc {
+	allContexts := "[\"TRACE_CONTEXT\",\"GRPC_BIN\",\"CLOUD_TRACE_CONTEXT\",\"B3\"]"
```
Comment (mandarjog) [inline pkg/bootstrap/option/convert.go:187]: Nit: use back quotes

Comment (brianwolfe) [reply pkg/bootstrap/option/convert.go]: Done

File: pkg/bootstrap/option/instances_test.go [hunk 4]
```diff
@@ -204,6 +204,37 @@ func TestOptions(t *testing.T) {
...
+		{
+			testName: "openCensusAgent order context",
+			key:      "openCensusAgentContexts",
+			option: option.OpenCensusAgentContexts([]meshAPI.Tracing_OpenCensusAgent_TraceContext{
+				meshAPI.Tracing_OpenCensusAgent_CLOUD_TRACE_CONTEXT,
+				meshAPI.Tracing_OpenCensusAgent_B3,
+				meshAPI.Tracing_OpenCensusAgent_GRPC_BIN,
+				meshAPI.Tracing_OpenCensusAgent_W3C_TRACE_CONTEXT,
+			}),
+			expected: "[\"CLOUD_TRACE_CONTEXT\",\"B3\",\"GRPC_BIN\",\"TRACE_CONTEXT\"]",
```
Comment (mandarjog) [inline pkg/bootstrap/option/instances_test.go:228]: Sorry missed this earlier, but you could use back quotes everywhere.

File: pkg/bootstrap/testdata/xdsproxy_golden.json [hunk 5]
```diff
@@ -249,9 +249,11 @@
       "transport_api_version": "V3",
       "grpc_services": [
         {
+        
           "envoy_grpc": {
             "cluster_name": "xds-grpc"
           }
+
```
Comment (mandarjog) [inline pkg/bootstrap/testdata/xdsproxy_golden.json:256]: Why is this change here?

**Review by mandarjog — COMMENTED** [review]
Can we add an e2e test here ?
Take a look At the stackdriver test that uses fake endpoint.
@gargnupur can you point him to it?

File: pkg/bootstrap/option/convert.go [hunk 6]
```diff
@@ -180,6 +181,33 @@ func durationConverter(value *types.Duration) convertFunc {
...
+			switch c {
+			// Ignore UNSPECIFIED
+			case meshAPI.Tracing_OpenCensusAgent_W3C_TRACE_CONTEXT:
+				envoyContexts = append(envoyContexts, "TRACE_CONTEXT")
+			case meshAPI.Tracing_OpenCensusAgent_GRPC_BIN:
+				envoyContexts = append(envoyContexts, "GRPC_BIN")
+			case meshAPI.Tracing_OpenCensusAgent_CLOUD_TRACE_CONTEXT:
+				envoyContexts = append(envoyContexts, "CLOUD_TRACE_CONTEXT")
+			case meshAPI.Tracing_OpenCensusAgent_B3:
+				envoyContexts = append(envoyContexts, "B3")
```
Comment (nrjpoddar) [inline pkg/bootstrap/option/convert.go:204]: Should there be a default error case in switch or by construction there cannot be values other than the ones you’re switching on?

File: releasenotes/notes/opencensusagent.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+apiVersion: release-notes/v2
+kind: feature
+area: telemetry
+
+releaseNotes:
+  - |
+    **Updated** added support for OpenCensusAgent formatted trace export with configurable trace context headers.
```
Comment (nrjpoddar) [inline releasenotes/notes/opencensusagent.yaml:7]: Should we add the new tracers that are now supported explicitly?

**Review by nrjpoddar — APPROVED** [review]

**Comment by gargnupur** [comment]
> Can we add an e2e test here ?
> Take a look At the stackdriver test that uses fake endpoint.
> @gargnupur can you point him to it?

Stackdriver tracing test: https://github.com/istio/istio/blob/master/tests/integration/telemetry/stackdriver/stackdriver_filter_test.go#L330:1
Zipkin tracing test:
https://github.com/istio/istio/blob/master/tests/integration/telemetry/tracing/servertracing/tracing_test.go

File: tools/packaging/common/envoy_bootstrap.json [hunk 8]
```diff
@@ -627,6 +627,29 @@
...
+        "@type": "type.googleapis.com/envoy.config.trace.v3.OpenCensusConfig",
+        "ocagent_exporter_enabled": true,
+        "ocagent_address": "{{ .openCensusAgent }}",
+        "incoming_trace_context": {{ .openCensusAgentContexts }},
+        "outgoing_trace_context": {{ .openCensusAgentContexts }},
+        "trace_config": {
+          "constant_sampler": {
+            "decision": "ALWAYS_PARENT"
+          },
+          "max_number_of_annotations": 200,
```
Comment (gargnupur) [inline tools/packaging/common/envoy_bootstrap.json:645]: should this be configurable?

File: pkg/bootstrap/option/convert.go (hunk 6 above)
Comment (brianwolfe) [reply pkg/bootstrap/option/convert.go:204]: By construction there cannot be more, aside from Tracing_OpenCensusAgent_UNSPECIFIED, which is intentionally skipped.

File: releasenotes/notes/opencensusagent.yaml (hunk 7 above)
Comment (brianwolfe) [reply releasenotes/notes/opencensusagent.yaml:7]: Do you mean the contexts that are supported? This only adds a single tracer w/ multiple contexts propagation header options.

File: tools/packaging/common/envoy_bootstrap.json (hunk 8 above)
Comment (brianwolfe) [reply tools/packaging/common/envoy_bootstrap.json:645]: @nrjpoddar and I discussed in the API PR and I ended up removing the tuning vars from the API. If you think it is worthwhile, I can add env variable tuning of these, but it doesn't appear to have a real use-case by users.

https://github.com/istio/api/pull/1591

File: pkg/bootstrap/testdata/xdsproxy_golden.json (hunk 5 above)
Comment (brianwolfe) [reply pkg/bootstrap/testdata/xdsproxy_golden.json:256]: This is unrelated. It looks like a previous change did not actually update using the standard mechanism of:
`REFRESH_GOLDEN=true go test ./pkg/bootstrap/...`
I can remove it, but it will need to be updated eventually anyway to make it easier to actually make changes to the template.

File: pkg/bootstrap/option/instances_test.go (hunk 4 above)
Comment (brianwolfe) [reply pkg/bootstrap/option/instances_test.go:228]: Done. You are right, I should have looked for all usage when you made that comment. Thanks!

**Comment by brianwolfe** [comment]
I'll work on adding the e2e test.

File: releasenotes/notes/opencensusagent.yaml (hunk 7 above)
Comment (nrjpoddar) [reply releasenotes/notes/opencensusagent.yaml:7]: I see. I think mentioning OpenCensus here and a doc for contexts make sense.

**Comment by brianwolfe** [comment]
FYI, first integration test attempt, so I ran into some issues. Got it going now though: https://github.com/istio/istio/issues/27061

File: pkg/bootstrap/testdata/xdsproxy_golden.json (hunk 5 above)
Comment (brianwolfe) [reply pkg/bootstrap/testdata/xdsproxy_golden.json:256]: I backed this change out, I'll submit seperately.

**Comment by brianwolfe** [comment]
@gargnupur Could you take a look at what I did here to try and test the OpenCensusAgent trace exporter? It's my first try at doing an e2e test in this framework. I think it makes sense, but I'm not positive.

I configured it to do: proxy -> opentelemetry-collector -> zipkin so we can reuse the zipkin querying for opencensusagent traces.

**Comment by brianwolfe** [comment]
@mandarjog is this similar to what you had in mind for an e2e test?

File: pkg/test/framework/components/opentelemetry/opentelemetry-collector.yaml [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: opentelemetry-collector
+  namespace: istio-system
```
Comment (bianpengyuan) [inline pkg/test/framework/components/opentelemetry/opentelemetry-collector.yaml:19]: Should remove istio-system namespace from here? The component logic decides which telemetry namespace this file is applied to.

**Review by bianpengyuan — COMMENTED** [review]
The new tracing test lgtm

**Review by bianpengyuan — APPROVED** [review]

File: releasenotes/notes/opencensusagent.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (brianwolfe) [reply releasenotes/notes/opencensusagent.yaml:7]: Done.

File: pkg/test/framework/components/opentelemetry/opentelemetry-collector.yaml (hunk 9 above)
Comment (brianwolfe) [reply pkg/test/framework/components/opentelemetry/opentelemetry-collector.yaml]: Done

**Comment by brianwolfe** [comment]
@howardjohn can you take a look for environments or let me know someone better to review?

**Review by howardjohn — APPROVED** [review]
